File: swift/HikeTimeCore/Sources/HikeTimeCore/GPX.swift
import Foundation

/// Разбор GPX: точки трека, высоты и время, если записаны.
public struct GPXTrack: Sendable {
    public let points: [GeoPoint]
    public let elevations: [Double]?     // только если есть у всех точек
    public let times: [Date]?            // только если есть у всех точек

    /// Полное время по меткам, часы.
    public var elapsedHours: Double? {
        guard let t = times, let a = t.first, let b = t.last else { return nil }
        return b.timeIntervalSince(a) / 3600
    }

    /// Время в движении: выбрасываем участки медленнее stopSpeed м/с.
    public func movingHours(stopSpeed: Double = 0.3) -> Double? {
        guard let t = times else { return nil }
        var total = 0.0
        for i in 1..<points.count {
            let dt = t[i].timeIntervalSince(t[i - 1])
            guard dt > 0 else { continue }
            if Geo.haversine(points[i - 1], points[i]) / dt >= stopSpeed {
                total += dt
            }
        }
        return total / 3600
    }
}

public enum GPX {

    public static func parse(_ data: Data) -> GPXTrack? {
        let d = Parser()
        let p = XMLParser(data: data)
        p.delegate = d
        guard p.parse(), !d.points.isEmpty else { return nil }
        return GPXTrack(
            points: d.points,
            elevations: d.eles.count == d.points.count ? d.eles : nil,
            times: d.times.count == d.points.count ? d.times : nil
        )
    }

    private final class Parser: NSObject, XMLParserDelegate {
        var points: [GeoPoint] = []
        var eles: [Double] = []
        var times: [Date] = []
        private var text = ""
        private var inPoint = false
        private let iso = ISO8601DateFormatter()

        func parser(_ parser: XMLParser, didStartElement name: String,
                    namespaceURI: String?, qualifiedName: String?,
                    attributes attrs: [String: String] = [:]) {
            let tag = name.lowercased()
            if tag == "trkpt" || tag == "rtept" {
                inPoint = true
                if let la = Double(attrs["lat"] ?? ""),
                   let lo = Double(attrs["lon"] ?? "") {
                    points.append(GeoPoint(lat: la, lon: lo))
                }
            }
            text = ""
        }

        func parser(_ parser: XMLParser, foundCharacters s: String) { text += s }

        func parser(_ parser: XMLParser, didEndElement name: String,
                    namespaceURI: String?, qualifiedName: String?) {
            let tag = name.lowercased()
            let value = text.trimmingCharacters(in: .whitespacesAndNewlines)
            if inPoint && tag == "ele", let v = Double(value) { eles.append(v) }
            if inPoint && tag == "time", let t = iso.date(from: value) { times.append(t) }
            if tag == "trkpt" || tag == "rtept" { inPoint = false }
        }
    }
}

File: model/gpx.py
"""Разбор GPX без внешних зависимостей."""
import xml.etree.ElementTree as ET
from datetime import datetime, timezone


def _text_time(s):
    s = s.strip().replace("Z", "+00:00")
    try:
        return datetime.fromisoformat(s).astimezone(timezone.utc)
    except ValueError:
        return None


def load(path):
    """-> dict: points [(lat,lon)], ele [м] или None, times [datetime] или None."""
    tree = ET.parse(path)
    root = tree.getroot()
    ns = {"g": root.tag.split("}")[0].strip("{")} if "}" in root.tag else {}
    q = (lambda t: f"g:{t}") if ns else (lambda t: t)

    pts, eles, times = [], [], []
    nodes = root.findall(f".//{q('trkpt')}", ns) or root.findall(f".//{q('rtept')}", ns)
    for p in nodes:
        pts.append((float(p.get("lat")), float(p.get("lon"))))
        e = p.find(q("ele"), ns)
        eles.append(float(e.text) if e is not None and e.text else None)
        t = p.find(q("time"), ns)
        times.append(_text_time(t.text) if t is not None and t.text else None)

    return {
        "points": pts,
        "ele": eles if all(e is not None for e in eles) and eles else None,
        "times": times if all(t is not None for t in times) and times else None,
    }


def elapsed_hours(times):
    if not times:
        return None
    return (times[-1] - times[0]).total_seconds() / 3600.0


def moving_hours(times, points, stop_speed=0.3):
    """Время в движении: выбрасываем участки медленнее stop_speed м/с (привалы)."""
    from .geo import haversine
    if not times:
        return None
    total = 0.0
    for i in range(1, len(times)):
        dt = (times[i] - times[i - 1]).total_seconds()
        if dt <= 0:
            continue
        d = haversine(points[i - 1], points[i])
        if d / dt >= stop_speed:
            total += dt
    return total / 3600.0
